### ISSUE TITLE
4 pictures in post

### ISSUE THREAD

**Issue by mkostera** [mastodon/mastodon#24461]
### Pitch

Mastodon should allow more pictures in post at least. Ideally without limitation. 

### Motivation

Different Fediverse software allow add more pictures to post, but it's not visible in Mastodon. This reduces the usability of Mastodon.

**Comment by joaoGabriel55** [comment]
I can work on a solution for that @mkostera!

**Comment by joaoGabriel55** [comment]
@mkostera is this the problem that you are facing? I think that would be good think in another limit number, cause allow the user send photos without limits can generate some side-effects. The Twitter case, only accept 4 images as Mastodon.

![image](https://user-images.githubusercontent.com/23138717/230984380-1e142641-962c-4079-bf90-b981bae872f6.png)

**Comment by joaoGabriel55** [comment]
What you think @Gargron?

**Comment by cdp1337** [comment]
I'm whipping up a solution for this on my personal fork, I set the arbitrary limit to 24 images just to not overwhelm the server too much, but still allow the user to select more images at a time.  I'm also "fixing" some Javascript by moving the presentation logic of image sizes and positions into the CSS _where it should have been in the first place_.

If Gargron is interested in this feature, I can put together a merge request once it's complete, otherwise it'll just live in my personal repo.

I'm not opposed to increasing the number to something higher than 24, but given how Mastodon is built, it's hard-coded in a bunch of places to expect _something_.  What'cha think?  24 good enough?  48?


![Working example of a single post with 14 images](https://user-images.githubusercontent.com/737865/232966857-dfb7fab1-d9a4-44ad-bd85-38e5a403fe84.png)

_Working (dev) screenshot of a post with 14 images attached_

**Comment by mkostera** [comment]
>. I'm also "fixing" some Javascript by moving the presentation logic of image sizes and positions into the CSS _where it should have been in the first place_.

If you mean ordering of picture, then PLEASE make merge request ASAP. It's really annoying bug.
https://github.com/mastodon/mastodon/issues/24462

**Comment by ismailghedamsi** [comment]
I really need this. Every day I post new images of hip-hop albums that dropped generally there is more than 4 albums that dropped and  on Friday it's over 20

> @mkostera I think that would be good think in another limit number, cause allow the user send photos without limits can generate some side-effects. The Twitter case, only accepts 4 images as Mastodon.
The more advantage mastodon have over Twitter the better. Also, Instagram and Facebook also allows 10 pictures.

**Comment by tejunk** [comment]
I would be great to see the solution.

**Comment by ileodo** [comment]
> I'm whipping up a solution for this on my personal fork, I set the arbitrary limit to 24 images just to not overwhelm the server too much, but still allow the user to select more images at a time. I'm also "fixing" some Javascript by moving the presentation logic of image sizes and positions into the CSS _where it should have been in the first place_.
> 
> If Gargron is interested in this feature, I can put together a merge request once it's complete, otherwise it'll just live in my personal repo.
> 
> I'm not opposed to increasing the number to something higher than 24, but given how Mastodon is built, it's hard-coded in a bunch of places to expect _something_. What'cha think? 24 good enough? 48?
> 
> ![Working example of a single post with 14 images](https://user-images.githubusercontent.com/737865/232966857-dfb7fab1-d9a4-44ad-bd85-38e5a403fe84.png)
> 
> _Working (dev) screenshot of a post with 14 images attached_

@cdp1337 great work! do you have a clean commit (or PR) for that changes in your personal fork?

**Comment by ileodo** [comment]
https://github.com/mastodon/mastodon/pull/29073 made this PR as a baby step.

**Comment by cdp1337** [comment]
> @cdp1337 great work! do you have a clean commit (or PR) for that changes in your personal fork?

I try to maintain branches with each individual feature I implement so others can cherry pick what they want to port.  My version was on the 4.1 codebase so it's out of date now, but working tonight to port it to latest 4.2 stable, (it'll be available under features/4.2-whatever-issue-number in my personal repo when I'm done).

**Comment by ileodo** [comment]
@cdp1337 thanks! I found that a few days back and made a https://github.com/mastodon/mastodon/pull/29073 PR to unify all the magic number 4 first.

**Comment by cdp1337** [comment]
Oh nice, I see you're using a configurable parameter in your version!  That's a good idea; I was contemplating doing that in my fork, but ultimately opted against unnecessary complications in the code.  This has been a feature request for at least 4 years, but Eugene has thus far refused to deviate away from a fixed 4-tile concept.

**Comment by ileodo** [comment]
I can feel that. from my point of view, even it has to be 4, this magic number has to be unified. (so people have different idea can easily config their instance differently).

also in the instance api end point, there is a field `max_media_attachement`, not sure if that means the community want it to be a configurable parameter.

**Comment by ileodo** [comment]
#29073 is passing CI and ready for review&merge

**Comment by janvlug** [comment]
I know that there are certain programs that allow for posting on several social media platforms. Apparently, currently the limit on Mastodon is still four pictures per post. This results in the fact that certain messages that can be posted on other social media platforms cannot be posted on Mastodon. This is a pity. What is the status of this issue, from the text above I get the impression that the corresponding MR is merged. Is there still other work to do? Or can instances already increase the number of allowed pictures in a post to above four?

**Comment by todoneunl** [comment]
I see three merged PR's. One from @ileodo (https://github.com/mastodon/mastodon/pull/29073), and (at least) two from @ClearlyClaire (https://github.com/mastodon/mastodon/pull/30932, https://github.com/mastodon/mastodon/pull/31616). Does this mean if we change MEDIA_ATTACHMENTS_LIMIT to a certain number on the server side (.env.production?) that our users can upload and display more images?  

I'm especially asking because we had political campaigns in the Netherlands last week and people using Buffer could not upload more than 4 images on Mastodon. Other micro-blog platforms like X/Twitter and Threads have no problem with this.

**Comment by this-stupid-web** [comment]
Same here, I'd like to recommend Mastodon as replacement to Facebook for public communications, but this solution is years behind, while they allow more than 15 images and videos per post Mastodon apparently only allows either 4 images or 1 video but not both at the same time. I can't suggest or recommend anyone to use Mastodon if it's worse than what they use now. Please don't limit the software, people want to use it to post and communicate

**Comment by mlo-3** [comment]
Yes, this is important in particular to repost Instagram posts, and FB ones. People posts the next images in comments to the post, which is annoying to use. 

Also the text size limit is too bad. Text is not so data heavy, should be far more caracters, and reduced with a "show full text" clic.

Thanks

**Comment by texttheater** [comment]
Mastodon should at least **warn** users when not all media are shown. With public posts, this warning could contain a link to the original post.

**Comment by todoneunl** [comment]
Meanwhile Bluesky allows 10 images. This is still not a priority it seems (not on roadmap). We also didn't receive a reply on this question: https://github.com/mastodon/mastodon/issues/24461#issuecomment-3476468162

**Comment by craigS89** [comment]
I really wish this issue would get some attention. Even if the Mastodon Team don't want to allow their own app to upload more than 4 images, it should still respect incoming posts and not strip out images.
hell, just a bandaid solution of letting users know the post is now missing images would help.
